src/nix/search.cc

```diff
@@ -181,15 +181,21 @@ struct CmdSearch : InstallableCommand, MixJSON
                 else if (initialRecurse)
                     recurse();
 
-                else if (attrPathS[0] == "legacyPackages" && attrPath.size() > 2) {
+                else if (file || expr || (attrPathS[0] == "legacyPackages" && attrPath.size() > 2))
```

Comment:
```suggestion
                else if (file || expr || (attrPath.size() > 2))
```
The "legacyPackages" restriction prevents ability to leverage the RecurseForDerivations logic in any other attribute path. This also helps match the behavior a few lines below, also removing the restriction.